### PR TITLE
feat!: Improve reliability, add unit tests

### PR DESCRIPTION
**Pull request by definitelyokay**
## Description

Refactors Alchemist to fix an issue with pending timers still being present when golden testing certain widgets. Alchemist was performing `ci` and `platform` golden tests inside of the same `testWidgets` call, which does not allow for pending timers to be flushed through the event loop on certain machines (or perhaps through the fakeasync machinery used under the hood by `flutter_test`). We noticed the problem on my Apple Silicon machine, but not on @Kirpal's intel mac.

After some investigation, we realized that using an individual `testWidgets` call for each golden test allows `flutter_test` to properly cleanup after a widget. We took advantage of `flutter_test`'s `TestVariant` system, which enable us to perform a `testWidgets` call for each platform that a golden test is configured to run on.

While making and testing this change, we refactored additional pieces of Alchemist so that it could be properly unit-tested. To preserve confidence of Alchemist's total behavior, we have created a new `smoke_tests` folder inside of `./test`, which can contain proper golden tests (which are effectively integration tests on Alchemist itself) and regression scenarios.

We have added our first regression scenario in `test/smoke_tests/timer_button_smoke_test.dart`:

```dart
import 'dart:async';

import 'package:alchemist/src/golden_test.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

class TimerButton extends StatelessWidget {
  const TimerButton({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ElevatedButton(
      child: const Text('Button'),
      onPressed: () {
        Timer(Duration.zero, () {});
      },
    );
  }
}

void main() {
  group('smoke test', () {
    goldenTest(
      'succeeds after tapping button with timer',
      fileName: 'timer_button_smoke_test',
      pumpBeforeTest: (tester) async {
        await tester.tap(find.byType(TimerButton));
        await tester.pumpAndSettle();
      },
      widget: const TimerButton(),
    );
  });
}
```

### Public API Changes

A few minor public API changes have been made.

- `TestComparisonPredicate` has been removed. We think everything it can do can be done with test tags, so we'd rather encourage the use of those since they're a vanilla part of `flutter_test`.
-  `PlatformGoldensConfig` and `CiGoldensConfig` share the same superclass, `GoldensConfig`. Doing so greatly simplified the `TestVariant` implementation.
- `FilePathResolver` now has a `FutureOr<String>` return type for convenience.

### Structural Changes

To enable Alchemist to be unit-tested, Alchemist's had to be uncoupled from `flutter_test` directly. Alchemist now consists of several layers that work together, each of which can be mocked, faked, or stubbed as needed for individual testing.

At the public level, `goldenTest` uses a `GoldenTestRunner` and `GoldenTestAdapter` to interface with `flutter_test`.

Throughout Alchemist, there are now various globals that allow for easy overriding in tests. These usually appear in little groups, like the `testWidgetsFn` override in `golden_test_adapter`:

```dart
/// Default testWidgets function that simply points to Flutter's [testWidgets].
TestWidgetsFn defaultTestWidgetsFn = testWidgets;
TestWidgetsFn _testWidgetsFn = defaultTestWidgetsFn;

/// A function to be used in place of `flutter_test`'s `testWidgets` function.
/// The ability to use a stub method here makes it easier to test Alchemist.
TestWidgetsFn get testWidgetsFn => _testWidgetsFn;
set testWidgetsFn(TestWidgetsFn value) => _testWidgetsFn = value;
```

This style of these global overrides is itself inspired by the source for `matchesGoldenFile`, which allows for its own `goldenFileComparator` to be overridden.

Likewise, the override functionality of `HostPlatform` was refactored to match this approach for the sake of consistency.

@jeroen-meijer If you have a chance, would love your insight on all this before we do anything crazy.

Special thanks to @Kirpal for creating the `AlchemistTestVariant` implementation and refactoring the configurations.

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [x] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [x] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [x] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: test/src/alchemist_config_test.dart [hunk 1]
```diff
@@ -222,52 +217,45 @@ void main() {
...
-        test('when platform is Linux', () {
-          setTestPlatform(HostPlatform.linux);
-
-          expect(
-            defaultValue.filePathResolver('foo'),
-            equals('goldens/linux/foo.png'),
-          );
-        });
+      tearDownAll(() {
+        currentHostPlatform = hostPlatform;
```
Comment (Kirpal) [inline test/src/alchemist_config_test.dart:247]:
```suggestion
        hostPlatform = currentHostPlatform;
```

File: lib/src/alchemist_config.dart [hunk 2]
```diff
@@ -239,115 +239,76 @@ class AlchemistConfig extends Equatable {
...
 
-  /// The default set of [platforms] that golden tests will run on.
-  ///
-  /// See [platforms] for more details.
-  static const _defaultPlatforms = HostPlatform.values;
+  /// Whether or not the golden tests should run.
+  final bool enabled;
 
-  /// The default value for the [enabled] field.
+  /// Whether of not all text should be rendered as colored boxes
```
Comment (Kirpal) [inline lib/src/alchemist_config.dart:284]:
```suggestion
  /// Whether or not all text should be rendered as colored boxes
```

File: test/src/alchemist_config_test.dart [hunk 3]
```diff
@@ -222,52 +217,45 @@ void main() {
...
+    group('environmentName', () {
+      var currentHostPlatform = hostPlatform;
+      late final HostPlatform nextHostPlatform;
+
+      setUpAll(() {
+        // Pick an environment that's different from the current one to ensure
+        // a proper test.
+        nextHostPlatform = currentHostPlatform == HostPlatform.linux
+            ? HostPlatform.macOS
+            : HostPlatform.linux;
```
Comment (Kirpal) [inline test/src/alchemist_config_test.dart:235]:
```suggestion
        nextHostPlatform = HostPlatform.values.firstWhere((platform) => currentHostPlatform != platform);
```

File: test/src/host_platform_test.dart [hunk 4]
```diff
@@ -7,6 +7,17 @@ import 'package:test/test.dart';
 class MockHostPlatform extends Mock implements HostPlatform {}
 
 void main() {
+  group('hostPlatform override', () {
+    test('overrides and returns the given value', () {
+      final nextHostPlatform = hostPlatform == HostPlatform.linux
+          ? HostPlatform.macOS
+          : HostPlatform.linux;
```
Comment (Kirpal) [inline test/src/host_platform_test.dart:14]:
```suggestion
      final nextHostPlatform = HostPlatform.values.firstWhere((platform) => platform != hostPlatform);
```

File: test/src/alchemist_test_variant_test.dart [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        final value = MockCiGoldensConfig();
+        const environmentName = 'TEST';
+        when(() => value.environmentName).thenReturn(environmentName);
+        await expectLater(
+          variant.setUp(value),
+          completes,
+        );
+        expect(variant.currentConfig, value);
+        expect(variant.describeValue(value), environmentName);
+      });
```
Comment (Kirpal) [inline test/src/alchemist_test_variant_test.dart:67]:
```suggestion
      test('setUp sets current value', () async {
        final value = MockCiGoldensConfig();
        await expectLater(
          variant.setUp(value),
          completes,
        );
        expect(variant.currentConfig, value);
      });

      test('describeValue returns environment name', () async {
        final value = MockCiGoldensConfig();
        const environmentName = 'TEST';
        when(() => value.environmentName).thenReturn(environmentName);
        expect(variant.describeValue(value), environmentName);
      });
```

File: test/src/golden_test_runner_test.dart [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+      setUp(() {
+        adapter = MockAdapter();
+        goldenTestAdapter = adapter;
+      });
+
+      test('overrides value', () {
+        expect(goldenTestAdapter, adapter);
+      });
+
+      tearDownAll(() {
```
Comment (Kirpal) [inline test/src/golden_test_runner_test.dart:38]:
```suggestion
      tearDown(() {
```

File: test/src/golden_test_test.dart [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+          goldenPath: 'myGoldenFile',
+          widget: widget,
+          forceUpdate: any(named: 'forceUpdate'),
+          obscureText: any(named: 'obscureText'),
+          textScaleFactor: any(named: 'textScaleFactor'),
+          theme: ciTheme,
+          pumpBeforeTest: any(named: 'pumpBeforeTest'),
+          whilePerforming: any(named: 'whilePerforming'),
+        ),
+      );
```
Comment (Kirpal) [inline test/src/golden_test_test.dart:164]:
```suggestion
      ).called(1);
```

File: test/src/golden_test_test.dart [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+      // Verify [GoldenTestRunner.run] was not called for the platform config.
+      verify(
```
Comment (Kirpal) [inline test/src/golden_test_test.dart:166]:
```suggestion
      verifyNever(
```

File: test/src/golden_test_adapter_test.dart [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+        testWidgetsFn = customTestWidgets;
+      });
+
+      test('overrides value', () {
+        expect(testWidgetsFn, customTestWidgets);
+      });
+
+      tearDownAll(() {
+        testWidgetsFn = defaultTestWidgetsFn;
+      });
```
Comment (Kirpal) [inline test/src/golden_test_adapter_test.dart:81]:
```suggestion
      test('overrides value', () {
        testWidgetsFn = customTestWidgets;
        expect(testWidgetsFn, customTestWidgets);
        testWidgetsFn = defaultTestWidgetsFn;
      });
```

File: test/src/golden_test_adapter_test.dart [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+        setUpFn = customSetUp;
+      });
+
+      test('overrides value', () {
+        expect(setUpFn, customSetUp);
+      });
+
+      tearDownAll(() {
+        setUpFn = defaultSetUpFn;
+      });
```
Comment (Kirpal) [inline test/src/golden_test_adapter_test.dart:97]:
```suggestion
      test('overrides value', () {
        setUpFn = customSetUp;
        expect(setUpFn, customSetUp);
        setUpFn = defaultSetUpFn;
      });
```

File: test/src/golden_test_adapter_test.dart [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+        tearDownFn = customTearDown;
+      });
+
+      test('overrides value', () {
+        expect(tearDownFn, customTearDown);
+      });
+
+      tearDownAll(() {
+        tearDownFn = defaultTearDownFn;
+      });
```
Comment (Kirpal) [inline test/src/golden_test_adapter_test.dart:113]:
```suggestion
      test('overrides value', () {
        tearDownFn = customTearDown;
        expect(tearDownFn, customTearDown);
        tearDownFn = defaultTearDownFn;
      });
```

File: test/src/golden_test_adapter_test.dart [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+              'message',
+              contains('matchesGoldenFile'),
+            ),
+          );
+        }
+      });
+
+      tearDownAll(() {
+        goldenFileExpectationFn = defaultGoldenFileExpectation;
+      });
```
Comment (Kirpal) [inline test/src/golden_test_adapter_test.dart:56]:
```suggestion
      test('overrides value', () {
        goldenFileExpectationFn = customExpectation;
        expect(goldenFileExpectationFn, customExpectation);
        goldenFileExpectationFn = defaultGoldenFileExpectation;
      });

      test('original value invokes matchesGoldenFile', () async {
        try {
          goldenFileExpectationFn(1, 2);
          fail('expected matchesGoldenFile to be invoked');
        } catch (e) {
          expect(
            e,
            isA<TestFailure>().having(
              (testFailure) => testFailure.message,
              'message',
              contains('matchesGoldenFile'),
            ),
          );
        }
      });
```

File: test/src/golden_test_adapter_test.dart [hunk 13]
```diff
@@ -0,0 +1,426 @@
...
+        verify(
+          () => paintingContext.paintSingleChild(
+            any(that: isA<RenderObject>()),
+          ),
+        );
+
+        expect(uiImage, isA<ui.Image>());
+      });
+
+      tearDownAll(() {
```
Comment (Kirpal) [inline test/src/golden_test_adapter_test.dart:207]:
```suggestion
      tearDown(() {
```

**Review by Kirpal — CHANGES_REQUESTED** [review]
Looks really good overall. Had a couple of suggestions. Also, are we sure we want to strip text packages by default? I'm thinking that might be a specific need for Betterment, so maybe we should remove that functionality and just strip the text package before passing in the theme.

**Comment by definitelyokay** [comment]
@Kirpal Thank you for taking over this while I'm out 🚀

**Pull request merged**